libraries/AP_Avoidance/AP_Avoidance.cpp

```diff
@@ -0,0 +1,598 @@
+        break;
+    case MAV_COLLISION_THREAT_LEVEL_LOW:
+        new_state = STATE_WARN;
+        action = _warn_action;
+        break;
+    case MAV_COLLISION_THREAT_LEVEL_HIGH:
+        new_state = STATE_FAIL;
+        action = _fail_action;
+        break;
+    default:
+        internal_error();
+        new_state = STATE_CLEAR;
+        action = MAV_COLLISION_ACTION_NONE;
+        break;
+    }
+
+    uint32_t now = AP_HAL::millis();
+    bool should_change_state = false;
+    do {
+        if (new_state == _old_state) {
+            break;
+        }
+        // do not transition back from a fail or warn state too quickly:
+        if (now - _last_state_change_ms < _state_recovery_hysteresis*1000) {
+            if (new_state == STATE_CLEAR ||
+                (new_state == STATE_WARN && _old_state == STATE_FAIL)) {
+                break;
+            }
+        }
+        should_change_state = true;
+    } while(0);
+
+    if (should_change_state) {
+
+        GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_WARNING, "AVOID: Moving from state (%s) to state (%s)", _state_names[_old_state], _state_names[new_state]);
+
+        _last_state_change_ms = now;
+
+        // we can change state but not change handler - for example,
+        // the user has specified the recovery handler for Fail->Warn
+        // as "continue fail behaviour".
+
+        // Note that it is entirely possible to transition from
+        // Fail->Clear; in this case we honour both _recovery_f and
+        // _recovery_w
+
+        bool change_handler = true;
+
+        if (_old_state == STATE_FAIL) {
+            // transitioning from fail -> warn
+            switch(_fail_recovery) {
+            case AVOIDANCE_RECOVERY_F_CONTINUE_FAIL:
+                change_handler = false;
+                break;
+            case AVOIDANCE_RECOVERY_F_MOVE_TO_WARN:
+                change_handler = true;
+                break;
+            }
+        }
+        if (new_state == STATE_CLEAR) {
+            // transitioning from (fail or warn) to clear
+            switch(_warn_recovery) {
+            case AVOIDANCE_RECOVERY_W_CONTINUE_ACTION:
+                change_handler = true;
+                break;
+            case AVOIDANCE_RECOVERY_W_RESUME:
+                change_handler = true;
+                break;
+            }
+        }
+
+        if (change_handler) {
+            AvoidanceHandler *new_avoidance_handler;
+            if (new_state == STATE_CLEAR) {
+                new_avoidance_handler = nullptr;
+            } else {
+                AvoidanceHandler &tmp = handler_for_action((MAV_COLLISION_ACTION)action);
+                new_avoidance_handler = &tmp;
+            }
+
+            if (new_avoidance_handler != _current_avoidance_handler) {
+                GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_WARNING, "AVOID: Moving to avoidance handler (%s)", ((new_avoidance_handler != nullptr) ? new_avoidance_handler->name() : "NULL"));
```

Comment:
I'm worried all these %s stuff may be too long. It won't hurt anything, still runs fine, but they get truncated